### PR TITLE
Added support for BYO_CRYPTO, still need to add CI to test, but we ne…

### PR DESCRIPTION
**Pull request by JonathanHenson**
…ed extra build steps.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -37,6 +37,8 @@ include(CTest)
 option(BUILD_RELOCATABLE_BINARIES
         "Build Relocatable Binaries, this will turn off features that will fail on older kernels than used for the build."
         OFF)
+option(BYO_CRYPTO "Don't build a tls implementation or link against a crypto interface. This feature is only for unix builds currently."
```
Comment (xomyakus) [inline CMakeLists.txt:40]: Should w name it BYO_CRYPTO or BYO_TLS?

Comment (JonathanHenson) [reply CMakeLists.txt:40]: I'm thinking of having an identical option in aws-c-cal so we can make sure the option controls both builds from a higher-level binding.

[the remark this answers is not in the document]
Comment (JonathanHenson) [reply CMakeLists.txt:40]: was planning on tying the aws-c-cal and aws-c-io flags together for build simplicity and minimize customer mistakes.

File: CMakeLists.txt [hunk 2]
```diff
@@ -142,27 +143,20 @@ elseif (CMAKE_SYSTEM_NAME STREQUAL "FreeBSD" OR CMAKE_SYSTEM_NAME STREQUAL "NetB
...
+    file(GLOB AWS_IO_TLS_SRC
+            "source/s2n/*.c"
+            )
+    aws_use_package(s2n)
+endif()
 
-    	aws_use_package(s2n)
-    endif ()
-endif ()
+if (BYO_CRYPTO AND (APPLE OR WIN32))
```
[the remark this answers is not in the document]
Comment (JonathanHenson) [reply CMakeLists.txt:157]: It's possible and a bad idea. There's no reason to do it on those platforms.

File: include/aws/io/tls_channel_handler.h [hunk 3]
```diff
@@ -195,6 +201,23 @@ enum aws_tls_negotiation_status {
     AWS_TLS_NEGOTIATION_STATUS_FAILURE
 };
 
+#ifdef BYO_CRYPTO
```
[the remark this answers is not in the document]
Comment (JonathanHenson) [reply include/aws/io/tls_channel_handler.h:204]: that's actually a really good point. I'll make it public.

File: include/aws/io/tls_channel_handler.h [hunk 4]
```diff
@@ -450,6 +479,23 @@ AWS_IO_API struct aws_channel_handler *aws_tls_server_handler_new(
     struct aws_tls_connection_options *options,
     struct aws_channel_slot *slot);
 
+#ifdef BYO_CRYPTO
+/**
+ * If using BYO_CRYPTO, you need to call this function prior to creating any client channels in the application.
+ */
+AWS_IO_API void aws_tls_client_handler_new_set_callback(aws_tls_handler_new_fn *client_handler_new_fn);
```
[the remark this answers is not in the document]
Comment (JonathanHenson) [reply include/aws/io/tls_channel_handler.h:486]: sure.

File: include/aws/io/tls_channel_handler.h [hunk 5]
```diff
@@ -195,6 +201,35 @@ enum aws_tls_negotiation_status {
     AWS_TLS_NEGOTIATION_STATUS_FAILURE
 };
 
+#ifdef BYO_CRYPTO
+/**
+ * Callback for createing a TLS handler. If you're using this you're using BYO_CRYPTO. This function should return
```
Comment (justinboswell) [inline include/aws/io/tls_channel_handler.h:206]: creating

File: include/aws/io/tls_channel_handler.h [hunk 6]
```diff
@@ -195,6 +201,35 @@ enum aws_tls_negotiation_status {
...
+ * a fully implemented aws_channel_handler instance for TLS. Note: the aws_tls_options passed to your
+ * aws_tls_handler_new_fn contains multiple callbacks. Namely: aws_tls_on_negotiation_result_fn. You are responsible for
+ * invoking this function when TLs session negotiation has completed.
```
Comment (justinboswell) [inline include/aws/io/tls_channel_handler.h:209]: TLS

File: tests/tls_handler_test.c [hunk 7]
```diff
@@ -1809,17 +1811,17 @@ static void s_import_cert(void *ctx) {
...
-#endif /* !AWS_OS_IOS */
+#    endif /* !AWS_OS_IOS */
 }
 
-#define NUM_PAIRS 1
+#    define NUM_PAIRS 1
 static int s_test_concurrent_cert_import(struct aws_allocator *allocator, void *ctx) {
     (void)ctx;
     /* temporarily disable this on apple until we can fix importing to be more robust */
     /* temporarily disable this on linux until we can make CRYPTO_zalloc behave and stop angering ASan */
```
Comment (justinboswell) [inline tests/tls_handler_test.c:1821]: Oh hey, we can probably re-enable this on linux again once this is in.

File: README.md [hunk 8]
```diff
@@ -230,7 +230,11 @@ BSD Variants | s2n
 Apple Devices | Security Framework/ Secure Transport. See https://developer.apple.com/documentation/security/secure_transport
 Windows | Secure Channel. See https://msdn.microsoft.com/en-us/library/windows/desktop/aa380123(v=vs.85).aspx
 
-In addition, you can always write your own handler around your favorite implementation and use that.
+In addition, you can always write your own handler around your favorite implementation and use that. To provide your own
```
Comment (justinboswell) [inline README.md:233]: Should probably call out that this is on Linux only.

**Review by justinboswell — APPROVED** [review]
Fix-n-ship!

**Pull request merged**